docs/weave-tutorial/03-codec.md

```diff
@@ -0,0 +1,219 @@
+2 critical points needs to be explained in state models and messages.
+
+Firstly, You must have noticed:
+
+```protobuf
+weave.Metadata metadata = 1;
+```
+
+`weave.Metadata` enables upgrading the code on chain without downtime. It is required if you want to use any [weave provided extensions](https://github.com/iov-one/weave/tree/master/x). [Migrations](https://github.com/iov-one/weave/tree/v0.20.0/migration) package will be explained in more detail in further sections.
+
+The second one is how the magic `ID` field works. This will be explained in [Models](weave-tutorial/04-models.md) section.
+
+## Message Definitions
+
+```protobuf
+// CreateOrderMsg will offer to sell some currency on an orderbook
+// at a given price.
+message CreateOrderMsg {
+  weave.Metadata metadata = 1;
+  // Trader is the Address that will pay the offer, and get the matches.
+  // Defaults to x.MainSigner() if left blank
+  bytes trader = 2 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+  // OrderBookID must support Offer.Ticker as one of the two sides,
+  // Which side this order will be, is automatically inferred
+  bytes order_book_id = 3 [(gogoproto.customname) = "OrderBookID"];
+  // Offer is how much will be paid
+  coin.Coin offer = 4;
+  // Price is how much is requested for each unit of the offer token
+  Amount price = 5;
+}
+```
+
+Noticed `(gogoproto.casttype)` and `(gogoproto.customname)`?
+
+```protobuf
+bytes trader = 2 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+```
+
+`(gogoproto.casttype)` indicates this field will be cast to `weave.Address`. Compiled codec.pb.go field:
+
+```go
+// Address of trader that created order (ad gets paid from it)
+    Trader      github_com_iov_one_weave.Address `protobuf:"bytes,3,opt,name=trader,proto3,casttype=github.com/iov-one/weave.Address" json:"trader,omitempty"`
+```
+
+```protobuf
+bytes order_book_id = 3 [(gogoproto.customname) = "OrderBookID"];
+```
+
+`(gogoproto.customname)` indicates this field will be named as `OrderBookID`. Compiled codec.pb.go field:
+
+```go
+OrderBookID []byte `protobuf:"bytes,3,opt,name=order_book_id,json=orderBookId,proto3" json:"order_book_id,omitempty"`
+```
+
+Without it we would get:
+
+```go
+OrderBookId []byte `protobuf:"bytes,3,opt,name=order_book_id,json=orderBookId,proto3" json:"order_book_id,omitempty"`
+```
+
+As you can see above we define necessary fields that will be used by `handler` to interact with application state.
+
+## Compiling Proto Files
+
+To compile protobuf files, you need to have the [protoc](https://github.com/google/protobuf#protocol-compiler-installation) binary installed and a language-specific translator (gogo-protobuf in this case). This can be a bit of a pain, especially the first time, so the default **weave-starter-kit** [Makefile](https://github.com/iov-one/weave-starter-kit/blob/master/Makefile) contains helpers for you.
```

Comment:
`Is the word 'helpers' correct here? Is it a technical term or do you mean something like 'helpful hints'?`

Reply:
It is a technical word